src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcherProvider.cs

```diff
@@ -16,37 +18,32 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
     [Export(typeof(FileChangeWatcherProvider))]
     internal sealed class FileChangeWatcherProvider
     {
-        private readonly TaskCompletionSource<IVsAsyncFileChangeEx> _fileChangeService = new(TaskCreationOptions.RunContinuationsAsynchronously);
-
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public FileChangeWatcherProvider(IThreadingContext threadingContext, [Import(typeof(SVsServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
+        public FileChangeWatcherProvider(
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            [Import(typeof(SVsServiceProvider))] Shell.IAsyncServiceProvider serviceProvider)
         {
-            // We do not want background work to implicitly block on the availability of the SVsFileChangeEx to avoid any deadlock risk,
-            // since the first fetch for a file watcher might end up happening on the background.
-            Watcher = new FileChangeWatcher(_fileChangeService.Task);
-
-            System.Threading.Tasks.Task.Run(async () =>
+            var fileChangeService = Task.Factory.StartNew(
+                async () =>
                 {
-                    await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(threadingContext.DisposalToken);
 
                     var fileChangeService = (IVsAsyncFileChangeEx?)await serviceProvider.GetServiceAsync(typeof(SVsFileChangeEx)).ConfigureAwait(true);
                     Assumes.Present(fileChangeService);
-                    _fileChangeService.SetResult(fileChangeService);
-                });
+                    return fileChangeService;
+                },
+                CancellationToken.None,
```

Comment:
odd to use threadingContext.Disposable token *in* the body of the lambda, but not pass that to StartNew.  indeed... that seems strictly wrong here to me.  if it isn't wrong, please doc why it's ok.  thanks!

Reply:
The two are semantically equivalent. I can add to the second location.

Reply:
please do.  :)

Reply:
➡️ Applied change